### DISCUSSION TITLE
Moor files with imports in different packages

### DISCUSSION THREAD

**Discussion by remcoanker** [simolus3/drift#1447] [Q&A] [answered]
I'm trying to split an application (and database) into multiple packages. It seems to work, except for imports.

- Package A defines a table in a .moor file and imports a type converter from another package.
- The database package imports Package A's .moor file.

When i run the generator, I get the message "Generated helper file not found."

The generator can't resolve imports in moor files defined in other packages, even if the database package has the converter as dependency and imports it.

When I move all the table definitions to the database package, the dao's of other packages can't use the type converters in generated code.

Is it possible to define tables in .moor files across packages?

**Comment by simolus3** [comment, marked as answer]
> When i run the generator, I get the message "Generated helper file not found."

Do both package A and the database package have a dependency on `moor` in their pubspec?

> When I move all the table definitions to the database package, the dao's of other packages can't use the type converters in generated code.

Is this because of missing imports? I think you'd have to manually add imports to the database in some cases, as the type converters are defined on table classes generated as a database part. #490 would improve things here.

**Comment by remcoanker** [comment]
Here is a [minimal example](https://github.com/remcoanker/moor_multi_package_database) of what I'm trying to do.

Running the generator in de database package generates a database without the converter.

**Reply by simolus3** [reply]
Well, this was a tricky one to figure out.

After a good hour of debugging `build_runner`, it turns out that your `database` package actually needs a (regular, non-dev!) dependency on `moor_generator`. After adding that dependency it worked for me. I'll ask around to check if this is the desired behavior here.

**Reply by remcoanker** [reply]
Thanks for looking into it.

Do you mean like  [this](https://github.com/remcoanker/moor_multi_package_database/commit/f3001334d99161ede07d5ce4d94464c6abcbcbe8)?

That still generates the database without the `JsonConverter`

**Reply by simolus3** [reply]
Sorry, I meant that `moor_generator` has to be a regular dependency of the `module` package.

This is because the build system needs to run a pre-processing step on the moor file in that package, and this step only runs for packages with dependencies on `moor_generator`. Dev-dependencies only work for the root package (`database` in this case).

**Reply by remcoanker** [reply]
That does work, thanks.
